rfd/0214-identifier-first-login.md

```diff
@@ -0,0 +1,193 @@
+      value: okta-auditor
+    - name: groups
+      roles:
+        - access
+      value: okta-dev
+  audience: https://<cluster-url>/v1/webapi/saml/acs/testconnector
+  user_matchers: ["^joe@foo.com$", ".*@goteleport\.com$"]
+  cert: ''
+  display: Okta
+  entity_descriptor: |
+    <?xml version="1.0" encoding="UTF-8"?>
+    <md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata" entityID="http://www.example.com/00000000000000000000">
+    ...
+    </md:EntityDescriptor>
+  entity_descriptor_url: ''
+  issuer: http://www.example.com/00000000000000000000
+  service_provider_issuer: https://teleport.com/v1/webapi/saml/acs/testconnector
+  signing_key_pair:
+    cert: |
+      -----BEGIN CERTIFICATE-----
+      ...
+      -----END CERTIFICATE-----
+    private_key: |
+      -----BEGIN RSA PRIVATE KEY-----
+      ...
+      -----END RSA PRIVATE KEY-----
+  sso: https://www.example.com/app/00000000000000000000/sso/saml
+version: v2
+```
+
+### Implementation
+
+Upon loading the Web UI, the server returns a web config which contains some details about the cluster
+including the Teleport edition, login methods and auth connectors (stored in `window['GRV_CONFIG']`).
+This will now also include the RegEx matchers per auth connector.
+
+All matching and filtering will be done on the client side. This is so that we can avoid having to
+introduce any additional calls to the backend once the user has provided a username. This is faster
+and more performant, and also reduces complexity since it means we won't need to add and maintain any new
+endpoints. Remembered usernames will be kept in `localStorage`.
+
+#### Proto changes
+
+`SAMLConnectorSpecV2`, `OIDCConnectorSpecV3`, and `GithubConnectorSpecV3` will need to be updated to include the `user_matchers` field.
+
+```proto
+// SAMLConnectorSpecV2 is a SAML connector specification.
+message SAMLConnectorSpecV2 {
+  ...
+  // UserMatchers is a set of RegEx patterns to narrow down which username(s) this auth connector should
+  // match for identifier-first login.
+  repeated string UserMatchers = 19 [(gogoproto.jsontag) = "user_matchers,omitempty"]
+}
+```
+
+```proto
+// OIDCConnectorSpecV3 is an OIDC connector specification.
+//
+// It specifies configuration for Open ID Connect compatible external
+// identity provider: https://openid.net/specs/openid-connect-core-1_0.html
+message OIDCConnectorSpecV3 {
+ ...
+  // UserMatchers is a set of RegEx patterns to narrow down which username(s) this auth connector should
+  // match for identifier-first login.
+  repeated string UserMatchers = 21 [(gogoproto.jsontag) = "user_matchers,omitempty"]
+}
+```
+
+```proto
+// GithubConnectorSpecV3 is a Github connector specification.
+message GithubConnectorSpecV3 {
+   ...
+  // UserMatchers is a set of RegEx patterns to narrow down which username(s) this auth connector should
+  // match for identifier-first login.
+  repeated string UserMatchers = 10 [(gogoproto.jsontag) = "user_matchers,omitempty"]
```

Comment:
Does this match against the GitHub username or what? The mappings make sense when you can match against user emails, but if people have arbitrary GitHub handles is this even useful?

Reply:
The GitHub username itself doesn't actually matter much aside from possibly causing confusion, because these matcher mappings are entirely on the Teleport side and any string can work if configured to. This would be useful if for example you allowed all your employees (ie. everyone whose email ended with "@foo.com") to log in with either Okta or GitHub, you'd have the ability to make GitHub show up as an option.

I added it just for consistency's sake, but I assume this is probably not going to be a common use-case and might just cause unnecessary confusion, so I'll remove it and only support `user_matchers` for SAML and OIDC.

Reply:
Oh, right.